### PR TITLE
feat: anchored EAGLE-3 training

### PR DESCRIPTION
**Pull request by Dogacel**
Adds DFlash style `training.eagle3_num_anchors`: the TTT unroll is evaluated at N sampled anchor positions per sample instead of at every supervised position, so each depth costs O(N) rather than O(sequence length). Off by default; the dense Eagle3 path is untouched.

**Why:** Cheaper and uniform length-sequence training. Also a long sequence can dominate the loss heavily, therefore anchor sampling ensures each sample contributes roughly equally, removing the bias for model to lean towards longer samples.

## How it works

Traditionally, EAGLE-3 trains with TTT rollout on every loss enabled token. The KV cache structure is shown below. This is expensive in terms of memory and runtime, especially for longer sequences.

image

DFlash has a different training method, it randomly samples "anchor tokens" and predicts N block continuations from each token. It achieves this by sorting each anchor token and create a bidirectional attention mask for each block prediction. This technique helps DFlash train stabily on longer sequences, since anchor token count doesn't change for longer sequences. Faster and potentially more stable training.

Screenshot 2026-08-30 at 3 20 06 PM

This PR combines both approaches, instead of doing TTT rollout on every token, we sample random anchor tokens to calculate loss from. Each token unsampled contributes through KV cache contributes anyway, so we don't need the full coverage to get similar results.

image

This means we calculate loss from significantly tokens, however results already show this doesn't hurt meaningfully. Below is an image of loss calculation coverage. The dark color means contributes more to the loss, since there is weight decay on same TTT (Same as DFlash)

image

Below is anchored, note that each token indirectly contributes via their KV cache anyway.

image

Since each token's weight is different, model can be trained for more epochs. In each epoch, data order will be random, so an un-seen token, or a seen but different TTT / weight token can contribute in the other epoch.

## Runtime

**Attention only (TTT=4, N=256, one GB300, ms / peak GB)**

| seq | dense TTT | anchored SDPA | anchored BlockMask | anchored score_mod |
|---|---|---|---|---|
| 4096 | 1.78 / 0.24 | 1.12 / 0.08 | **0.97 / 0.08** | 0.97 / 0.08 |
| 6144 | 3.82 / 0.44 | **1.10 / 0.12** | 1.38 / 0.15 | 1.39 / 0.11 |
| 16384 | 20.77 / 1.11 | **2.30 / 0.30** | 3.44 / 0.34 | 3.45 / 0.31 |
| 32768 | 75.95 / 2.06 | **4.24 / 0.59** | 6.73 / 0.61 | 6.75 / 0.57 |
| 131072 | 1134.53 / 7.89 | **15.88 / 2.33** | 26.43 / 2.22 | 26.45 / 2.19 |

Dense time and memory grows quadratically, anchored is roughly flat. SDPA does **not** cost more memory, since the mask is built with `H=1` and broadcast and SDPA tiles rather than materialising the score matrix.

**End-to-end training (Qwen3-8B, 20K open-perfectblend, 5 epochs)**

Open-perfectblend doesn't have very long rows, so the benefits are minimal, but still visible.

image

DFlash is less data efficient, because predicted blocks are not overlapped with each other, so a token at position X is always treated as a token at (X mod block_size) position. Because position of a token also changes prediction hardness + loss weight, model doesn't see all tokens truly during training.

This has been validated by running EAGLE with non-overlapped anchors, the result was significantly worse than regular EAGLE, but it was better than DFlash and DFlash 2.

| Method | acc len | step/s | samples/s | opt steps | accum | state |
|---|---|---|---|---|---|---|
| DFlash | **0.39** | 5.1 | 10.1 | 12370 | 2 | completed |
| DFlash2 | **0.43** | 4.7 | 9.4 | 12370 | 2 | completed |
| EAGLE-3 (dense) | **0.78** | 17.4 | 17.4 | 24745 | 1 | completed |
| Anchored, scattered | **0.47** | 18.8 | 18.8 | 12163 | 1 | **died @ 12163** |
| Anchored, adjacent | **0.78** | 9.4 | 18.9 | 12370 | 2 | completed |

Also added block mask compilation for DFlash, which speeds up training by 10%,
```
block mask build is EAGER      →  9.26 samples/s
block mask build is COMPILED   → 10.24 samples/s   (1.106x)
```

## Multi-layer sweep (N=1024, TTT=4, one GB300, ms / peak GB)

| seq | | dense TTT | anchored SDPA | speedup |
|---|---|---|---|---|
| **4096** | L=1 | 1.79 / 0.24 | 1.24 / 0.17 | 1.44× |
| | L=2 | 3.53 / 0.24 | 2.84 / 0.17 | 1.24× |
| | L=5 | 8.78 / 0.24 | 6.99 / 0.17 | 1.26× |
| **6144** | L=1 | 3.82 / 0.44 | 1.44 / 0.23 | 2.65× |
| | L=2 | 7.63 / 0.44 | 4.51 / 0.29 | 1.69× |
| | L=5 | 18.99 / 0.44 | 11.15 / 0.29 | 1.70× |
| **32768** | L=1 | 75.89 / 2.06 | 5.37 / 1.02 | **14.1×** |
| | L=2 | 151.79 / 2.06 | 47.40 / 1.24 | 3.2× |
| | L=5 | 379.35 / 2.06 | 118.42 / 1.24 | 3.2× |

## MLA

| seq | dense TTT | anchored SDPA | speedup |
|---|---|---|---|
| 4096 | 2.89 / 0.92 | 1.21 / 0.43 | 2.4x |
| 6144 | 5.99 / 1.69 | 1.19 / 0.63 | 5.0x |
| 16384 | 34.27 / 4.20 | 2.17 / 1.60 | 15.8x |
| 32768 | 127.48 / 7.91 | 4.02 / 3.17 | **31.7x** |

Real end to end run timees.

| pblong10k, MLA draft, 16k seq | compute_time | data_time | step_time | samples/s |
|---|---|---|---|---|
| dense, 1 train / 3 infer | 0.1540 | 0.0374 | 0.1625 | 12.31 |
| anchored, 1 train / 3 infer | **0.1244** | 0.0299 | 0.1547 | 12.93 |

| data | median supervised | samples > N | GPU split | compute | end-to-end |
|---|---|---|---|---|---|
| pblong10k | 1401 | 100% | 1 train / 3 infer | **1.24x** | 1.05x |

### CODE REVIEW

File: torchspec/models/anchored_eagle3.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        cos, sin = cos.to(device), sin.to(device)
+        context_positions = torch.arange(seq_len, device=device).expand(bsz, -1)
+        key_cache, _ = apply_rotary_pos_emb(key_cache, key_cache, cos, sin, context_positions)
+
+        current = _gather(projected, anchors)
+
+        plosses, vlosses, acces, acc_counts, alphas = [], [], [], [], []
+        for depth in range(self.length):
+            positions = (anchors + depth).clamp(max=seq_len - 1)
+            fused = self._fuse(layer, _gather(embeddings, positions), current)
```
[the remark this answers is not in the document]
Comment (Dogacel) [reply torchspec/models/anchored_eagle3.py:237]: I think this reduces train/test time mismatch, so it is intentional and good to ship.

File: torchspec/models/anchored_eagle3.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                value_cache,
+                attn_mask=anchored_bool_mask(
+                    anchors, keep_mask, seq_len, num_anchors, key_cache.shape[2]
+                ),
+                enable_gqa=True,
+            )
+            attn_out = attn_out.transpose(1, 2).reshape(bsz, num_anchors, -1)
+
+            hidden = current + attn.o_proj(attn_out)
+            hidden = hidden + layer.mlp(layer.post_attention_layernorm(hidden))
```
[the remark this answers is not in the document]
Comment (Dogacel) [reply torchspec/models/anchored_eagle3.py:260]: Anchoring already saves significant memory, so not a P0.

**Review by yubofredwang — APPROVED** [review]
